tests/string_test.c

```diff
@@ -0,0 +1,366 @@
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char(&to_split, ';', &output));
+    ASSERT_INT_EQUALS(4, aws_array_list_length(&output));
+
+    struct aws_byte_cursor value = {0};
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 0));
+
+    char *expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 1));
+    expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 3));
+    expected = "testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_adj_tokens, test_char_split_adj_tokens_fn)
+
+static int test_char_split_with_max_splits_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = ";testa;testb;testc";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char_n(&to_split, ';', &output, 2));
+    ASSERT_INT_EQUALS(3, aws_array_list_length(&output));
+
+    struct aws_byte_cursor value = { 0 };
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 0));
+    char *expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 1));
+    expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb;testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_with_max_splits, test_char_split_with_max_splits_fn)
+
+static int test_char_split_output_too_small_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = "testa;testb;testc;";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    struct aws_byte_buf output_array[3] = {{0}};
+    ASSERT_SUCCESS(aws_array_list_init_static(&output, output_array, 3, sizeof(struct aws_byte_cursor)));
+    ASSERT_ERROR(AWS_ERROR_LIST_EXCEEDS_MAX_SIZE, aws_string_split_on_char(&to_split, ';', &output));
+   
```

Comment:
Should we assert on what is in the array now?

Reply:
good idea.